### ISSUE TITLE
Voice/Video chat beta: Missing Echo Cancellation

### ISSUE THREAD

**Issue by Spindel** [signalapp/Signal-Android#6241]
I have:
- [x] searched open and closed issues for duplicates
- [x] read https://github.com/WhisperSystems/Signal-Android/wiki/Submitting-useful-bug-reports

----------------------------------------

### Bug description
When doing an audio call on Speaker mode in the new Voice/Video call beta enabled, you can hear your own voice clearly reflected through the microphone.  This causes a rather high level of annoyance as you hear your own voice, with a double round-trip time in between, making it a very sad communication companion that constantly interrupts you when you speak.

### Steps to reproduce
- Enable voice calling beta
- Put both phones on speaker mode
- Pull up volume a bit
- Speak slowly and hear your own voice with a delay, routed through the aether.

**Actual result:**
Annoyingly noise of your own voice

**Expected result:** Effective echo cancellation on the remote side that cancels out the noises from the speaker against the microphone.

### Screenshots
Not applicable

### Device info
**Device 1**: Fairphone 2
**Device 2** Nexus 5x 
**Android version:** 7.1.1 and the current fairphone (5/6 something?)
**Signal version:** Pass.

### Link to debug log
Not applicable.

**Comment by haffenloher** [comment]
Do you observe this on both ends?

Just to be sure: are the devices isolated from each other / in different rooms?

**Comment by moxie0** [comment]
Some devices have broken hardware AEC. If you hear echo, the problem is on the *other* end (that device isn't canceling out the speaker audio from the microphone input), so we'd need information about the device on the other side.

**Comment by Spindel** [comment]
It was observed on one end, devices are isolated from each other with a nice wide swath of ocean and some land.

The device that would be missing AEC would be the Fairphone2, which is also, sadly, the one that I don't have here to get specifications & logs from.

**Comment by FeuRenard** [comment]
I have a Fairphone 2. I will try to reproduce this issue in the next few days.

**Comment by FeuRenard** [comment]
Just tested it. I heard a very clear echo speaking into device XU -> FP's echo cancellation failed. I heard the echo cancellation working speaking into device FP -> XU's echo cancellation works fine. There is always some soft noise coming from FP's earphone but directly after I spoke a word into FP this noise was somehow disturbed. I interpreted this as working echo cancellation.

tl;dr I can reproduce this issue and it's device FP's fault.

### Device info FP

**Device:** Fairphone 2
**Android version:** 5.1.1
**Signal version:** 3.29.6

### Device info XU

**Device:** Sony Xperia U
**Android version:** 4.4.4
**Signal version:** 3.29.6

**Comment by Spindel** [comment]
Thank you for verifying this. Can you follow up with upstream fairphone?

**Comment by FeuRenard** [comment]
I tested the dialer app with my landline phone and I hear no echo speaking to my landline phone. But maybe the landline phone can also do echo cancellation for the other side...

> Can you follow up with upstream fairphone?

I just revived this thread: https://forum.fairphone.com/t/echo-with-google-hangout/15745

**Comment by moxie0** [comment]
@FeuRenard I can try blacklisting the fairphone from hardware AEC in the webrtc code, can you give me the fairphone2's "ro.product.model" string?

**Comment by FeuRenard** [comment]
In settings -> about phone -> model number it says "FP2". Is this what you want? If not: From where can I get this string?

**Comment by moxie0** [comment]
Yeah I think that's it

**Comment by moxie0** [comment]
@FeuRenard can you try [this test apk](https://mega.nz/#!FtVDyDyQ!emuS0Hq0jXZDN_s5vYP8A-_A0tN-BwPGjvrQdJY3Qyk) on your FP2 and see if that helps?

**Comment by FeuRenard** [comment]
I tried. Good news: echo cancellation seems to work with this apk. Bad news: Fairphone 2 reboots after a couple of seconds.

**Comment by moxie0** [comment]
Hah, what!? Just... reboots?

**Comment by FeuRenard** [comment]
Yes. 4 out of 5 times I tried it rebooted. I sent you logcat of the final seconds before reboot via e-mail if you are interested.

**Comment by rainerzufall** [comment]
@FeuRenard can you explain how to install this apk (in the community forum for example)? Can i run it alongside my normal signal installation? Then I could try and test it on my Fairphone 2 tomorrow...

**Comment by moxie0** [comment]
Sounds like the fairphone has some issues. Either they need to fix the AEC or they need to fix the crash, I don't think there's anything else I can do.

**Comment by Trolldemorted** [comment]
My redmi note 3 pro is displaying the same issues: my conversation partner is complaining about echo, but on my end everything is fine.

If you push your commits from your testapk somewhere i can give it a shot.

`getprob` says:

```
[ro.product.board]: [MSM8952]
[ro.product.brand]: [Xiaomi]
[ro.product.cpu.abi]: [arm64-v8a]
[ro.product.cpu.abilist]: [arm64-v8a,armeabi-v7a,armeabi]
[ro.product.cpu.abilist32]: [armeabi-v7a,armeabi]
[ro.product.cpu.abilist64]: [arm64-v8a]
[ro.product.device]: [kenzo]
[ro.product.locale]: [en-US]
[ro.product.manufacturer]: [Xiaomi]
[ro.product.model]: [Redmi Note 3]
[ro.product.name]: [cm_kenzo]
```

Despite it saying redmi note 3 it is a redmi note 3 pro, which has slightly different hardware.

**Comment by FeuRenard** [comment]
Just for the record:
My conversation partner even complains about hearing herself while none of us uses the speaker mode. She uses an iPhone SE and I use a Fairphone 2.

**Comment by haffenloher** [comment]
@FeuRenard interesting. Sounds just like something I experienced when calling a Xiaomi Mi4 (ro.product.model = "MI 4LTE"):

1. I heard a loud and clear echo with a delay of maybe 1-2 seconds, regardless of whether the speaker setting on the Mi4 was on or off, which was very disturbing.
1. The first words of every sentence I spoke could be heard clearly on the Mi4, but (roughly) as soon as I started hearing the echo, my voice started to drop out on the Mi4. To be understood, I had to stop every few words and wait till the echo subsided.

We tried every combination of the three blacklists in [WebRtcAudioUtils.java](https://chromium.googlesource.com/external/webrtc/+/master/webrtc/modules/audio_device/android/java/src/org/webrtc/voiceengine/WebRtcAudioUtils.java#30). The minimal change required to make it work properly was to add "MI 4LTE" to both the `OPEN_SL_ES` and the `AEC` blacklists. Tested in a 40 minute video call with speaker mode on and off, no problems.

If you're interested in investigating this further, I could provide you with test builds of the AppRTC demo app with and without your model blacklisted.

**Comment by skulumani** [comment]
I've noticed similar echo issues with a Samsung S5? Is there any similarity between the hardware on these devices?

**Comment by haffenloher** [comment]
@skulumani what exactly are you seeing? You're hearing an echo when you do a WebRTC call to someone who has a Samsung Galaxy S5 and they enable speaker mode?

**Comment by Trolldemorted** [comment]
ftr: i too was not using the speaker mode, but the other party complained about echo.

**Comment by skulumani** [comment]
I (Moto X 2013) call a friend (S5) and I experienced an echo of my own
voice.

We're both using the new beta calling feature and from this thread it seems
to be a function of the S5 rather than my moto x.

Another thing is that the echo is almost the same volume as the incoming
audio. I suppose it's a function of the volume of the speaker of the S5.

We're both using the standard ear piece rather than the speakerphone. Only
I hear the echo while the friend does not.  I can do more testing and
submit logs if it'll be useful.

Is there any more information that might be useful. For example, I can do
some research into the hardware of the Samsung and Fairphone. I can get
debug logs or try and use the test apk from above.

**Comment by jeremymasters** [comment]
By the way @moxie0, I have an s5 I could test on if needed.

**Comment by swltr** [comment]
We're hearing echoes since video calls were added, too. (When video calls were in opt-in beta, we enabled it for testing, heard the echoes and disabled it again.) Before, we didn't hear any echoes at all.

Both devices are Wileyfox Swift (and the echoes can be heard on both ends).

**Comment by 0o-o0** [comment]
I am still getting echoes (Pixel Phone to Pixel Phone). One thing that I noticed is that I have speaker on when the video call starts. If I turn speaker off it mitigates the echoing (but doesn't make it go away). With speaker off it is still loud enough for me to hear the other person (in a quiet room). Not sure if it's being tested with/without speaker.

**Comment by moxie0** [comment]
Seems like we should get more aggressive about adding devices to the hardware AEC blacklist. @haffenloher I'm surprised that OpenSL was a contributing factor with echo.

**Comment by moxie0** [comment]
Pixel, S5, Redmi Note 3, and Xiaomi Mi4 should be taken care of in 4.1.0

**Comment by haffenloher** [comment]
@moxie0 i know, it's really weird. But then again, the symptoms were really weird too (super-strong and clear echo even with headphones).

Anyway, cool that you're adding more phones now! I did some tests with an echo-producing Samsung **Galaxy S4 Mini** and can confirm that adding "GT-I9195" to the hardware AEC blacklist fixes the echo.
Also noticed echo with a Sony "**Xperia SP**" (apparently that's the model string), but didn't have the chance to test blacklisting it yet.

**Comment by moxie0** [comment]
This is kind of crazy, are there any devices that *do* have functional hardware AEC?

**Comment by haffenloher** [comment]
So far I only saw it working with three devices :/
a pattern I noticed is that all the people I had echo problems with were running CM / LineageOS. Could that make a difference? Do any CM or Lineage users have working echo cancellation?

**Comment by moxie0** [comment]
Oh yeah that could definitely make a difference. I'd only want to make changes to the AEC blacklist if it's a documented problem on stock.

**Comment by flungo** [comment]
I have experienced this problem with the recipient having an **LG G3** running stock Android. They do not report that they can hear an echo the other way round.

**Comment by Trolldemorted** [comment]
I have had a webrtc-call with a different contact yesterday (i am still running 3.30.1), and the other party did not hear any echo. Are you really sure this is always only about hardware AEC on my end?

**Comment by moxie0** [comment]
It looks like we can make these changes at the application level, so I've reverted 4.1.0 back to stock M57 and added the blacklist info into the initialization logic. If anyone with an affected device can confirm these application-level changes are still effective that'd be great.

I also added the LG G3 here.

**Comment by McLoo** [comment]
Can't do the blacklist test, 
but just called a Galaxy S7 (stock OS) with massive echoing. 
My Sony Z5 Compact behaves nice and echoless

**Comment by flungo** [comment]
Would it be worth adding an advanced settings option to force software based AEC (with a suggestion to report if it fixes the echo problems in the description) to make testing which models require this easier?

**Comment by moxie0** [comment]
@flungo Might be a good idea but most people won't understand that the setting should be toggled by the person they're calling, not the person that hears the echo.

**Comment by moxie0** [comment]
Also, I added Samsung Galaxy S7 and Sony Xperisa SP to the blacklist.

**Comment by jeremymasters** [comment]
Should hardware AEC just be shut off to avoid keeping a blacklist until time for further research?

**Comment by lantsem** [comment]
Tested several android phones with Signal 4.1.0:

Motorola Moto X 2014 (XT1095), stock 6.0 - OK
Motorola Moto X 2014 (XT1095), LineageOS 14.1 (7.1.1) - echo
Wileyfox Swift, stock CM13 (6.0.1) - echo

@moxie0 We definately need at least an advanced option to manually enable software based AEC. But forcing software AEC for all phones looks like the simplest solution for compatibility.

**Comment by jp8** [comment]
On my device (Honor X5 on lineageos) I notice at the end of /system/etc/audio_effects.conf the following message:

#
# TODO: add default audio pre processor configurations after debug and tuning phase
#

And I suppose if I add something like the following to this file, I might enable the hardware acoustic echo cancelation..


pre_processing {
  voice_communication {
    aec {}
  }
}

**Comment by 0o-o0** [comment]
After update Pixel to Pixel seems good to go. Thank you! @moxie0 

(These devices are running stock android btw.)

**Comment by paride** [comment]
There: #6465. It seems that the hardware AEC is mostly broken, it should probably be enabled on a whitelist basis, or just disabled.

**Comment by lantsem** [comment]
I think that static blacklist/whitelist solutions isn't right for this problem. It would work with iOS, where you have only several supported phones, which can easily be tested and re-tested after major OS updates. But with Android we have hundreds phones, many of them receive major OS updates and needs to be re-tested (becase major update easily can break or fix something), and some of them may be used with custom software (which similarly can break or fix something).

For example my situation: I have two Moto X 2014, one is used with stock software with working hardware AEC (rare case), and another with custom software with broken hardware AEC. I want to use Signal for voice calls between them. How could be this possible with only static blacklist/whitelist? This phones has the same product model string (XT1095).

Probably you want to have working Signal on all android phones with all their diversity and not only on dozen "compatible" ones, which is listed in blacklist, right? BTW, you can hear echo even without speaker mode with normal voice calls. So this regression really hinders the use of calls in Signal.

Possible solutions:
1. Leave hardware AEC, continue to blacklisting phones tested and re-tested with latest stock software and make advanced option to disable hardware AEC for users with undestandable description (something like "enable only if the other party complain to you about echo of their own voice they hear").
2. Disable hardware AEC for everyone. Really simple and completely eliminates problem for all cases.
3. Disable hardware AEC for everyone and make advanced option to enable hardware AEC. But this option is risky, so it need to be hidden like android's settings for developers.

Personally I can use any of the three proposed solutions. But taking into account the diversity of android phones and the efforts required to maintain blacklist, I think the last two solutions is better for this case.

@moxie0 What do you think?

**Comment by jp8** [comment]
@lantsem 

I agree your option 1 seems unworkable, but I suppose a whitelist would be manageable.

In the longer term I can imagine a fourth possible solution in which when echo or feedback is detected (automatically) at one end of the conversation, the client sends a request for the other side to switch to software AEC.

**Comment by sigenc** [comment]
Nougat-MR2 (7.1.2) was released. This should provide a solution for Nexus 6P.

**Comment by xenithorb** [comment]
I believe I'm being hit by this between two bacon phones (OnePlus One) in a worse way than reported, not only does it echo, but when video is triggered it causes the typical screeching loopback noise as when you have two phones next to each other with speaker on. This happens regardless of the proximity of the phones (meaning that I confirmed that it happens when they are not next to each other and are too far away to be caused by loopback interference). Unfortunately now that the new WebRTC calling is out of beta and the old method deprecated, I can't really effectively privately call the only person I use private calling for. :(

**Comment by lantsem** [comment]
@jp8  
Yes, dynamic adjustment looks like possible solution for this problem. And webrtc M57 already has audio processing component that detects echo.

BTW, I remembered interesting fact. On March 15 I tested new video calling with Signal 3.31.4 on both ends between my Moto X2 (XT1095, custom 7.1.1) and Galaxy S5 (SM-G900F, stock 6.0.1). Except #6356 it was perfect, no echo at all on both ends. Next time I called on March 29, with Signal 4.1.0 on both ends (in which S5 was blacklisted) and other side clearly heard the echo from me in that time. Since then I have always been complaining about the echo. How could this be?

**Comment by haffenloher** [comment]
@lantsem if the person hearing the echo is the one using the Galaxy S5, then it's unrelated to the S5 being blacklisted

**Comment by lantsem** [comment]
@haffenloher I am fully aware of this and talking about different things:
1. S5 in my test didn't have an echo with 3.31.4 (before it was blacklisted).
2. My Moto X2 didn't have an echo with 3.31.4.
3. My Moto X2 got echo with 4.1.0 and have it in current version.
How could this be?

**Comment by haffenloher** [comment]
@lantsem I guess you had the X2's speaker enabled in all cases? It's possible that the (hardware) AEC performance differs depending on the room, where the phone lies, the speaker volume etc.

**Comment by lantsem** [comment]
We tested video with 3.31.4, so speaker was enabled. Now I'm getting complaints even without speaker and I hear an echo without speaker too (from Wileyfox Swift as I wrote earlier). Room condition is the same, phone software wasn't changed, volume too. Strangely.

**Comment by haffenloher** [comment]
> Now I'm getting complaints even without speaker

That's an issue I've seen exclusively with CM / LineageOS users so far. Looks like there's something seriously wrong on their end 😕

**Comment by lantsem** [comment]
Maybe something CAF related? I have an old Galaxy Nexus with LOS13, that seems to works without echo (it has some different sound artifact, not so important), but I'm not sure that there is any hardware AEC in it at all.

BTW, normal GSM voice calls works great on all this phones, no echo at all. All phones uses software AEC for GSM?

**Comment by moxie0** [comment]
> 2. Disable hardware AEC for everyone. Really simple and completely eliminates problem for all cases.

I'm leaning towards this, but it might be an overstatement to say that it will completely eliminate the problem. There could be a class of devices where hardware AEC works but software AEC doesn't for some reason, we don't know.

**Comment by lantsem** [comment]
Yes, I had to insert 'maybe' into this sentence. That's why I suggested a 3rd choise with hidden option to enable hardware AEC to overcome potentional problems and report back to developers. This option for toggle hardware/software mode can also be used for quick testing of different modes in the same conditions, so it could give back much more detailed feedback from advanced users.

But how bad it could be with global software AEC rollout? If you look at current blacklist and take into account the popularity of listed devices, hardware AEC issue already hit quite hard. Even S5, S7 and Google's own Nexus/Pixel (i.e. phones with good and long software support) do not garantee the absence of problems. And it could be much worse on many other cheaper phones. Please correct me, if I wrong, did redphone code use software AEC? Did you have a lot of reports about the AEC in that time? Personally I had no problems with redphone or SIP AOSP client (that seems to use software AEC). So the question is only about quality of software AEC in webrtc, am I right?

From the side it seems that the amount of phones, which could have problems with webrtc's software AEC, can be significantly lower than in the case of hardware AEC. Of course I can be wrong, and only real statistics will show.

**Comment by OfficiallyAhmad** [comment]
Can someone can do a pr to add Xperia Z to AEC blacklist.

ro.product.model : **C6603**

**Comment by paride** [comment]
@moxie0 my impression is that the hardware AEC is very often broken, even on high end devices with stock system images. Disabling the hardware AEC by default seems the most reasonable thing to do. If there are devices where software AEC doesn't work but hardware AEC does, they could be whitelisted, inverting the current logic.

**Comment by moxie0** [comment]
I've switched to a whitelist, so let's start this process over again. =) Any devices we know *do* work?

**Comment by AlfonsoMuskedunder** [comment]
I haven't had any problems with echos on my Moto G 2014 (Titan) on LineageOS 14.1, I can tell that much. Same for my partner's Moto G (Falcon) (on LineageOS 14.1 as well).

**Comment by RiseT** [comment]
> Any devices we know do work?

Wouldn't the appropriate question be...

**Any devices we know that don't work with software AEC?**

**Comment by moxie0** [comment]
@RiseT My intuition is that software AEC will always reduce performance in some way (battery life at the least), so we should try to use hardware AEC whenever we can.

**Comment by braaccckkk** [comment]
I didn't had any problems with my HTC 10 (LineageOS 14.1):
[ro.product.model]: [HTC 10]

**Comment by johanw666** [comment]
Sony Xperia Z1 Compact (D5503) and Z3 Compact (D5803) appear to work in the current hardware AEC setup.

**Comment by RiseT** [comment]
It looks like a lot of work keeping track of all the different combinations of device and OS version, though. Especially when just one (future) minor OS update on a device can break a previously working (and potentially whitelisted) hardware AEC.

**Comment by RiseT** [comment]
Another thought:
How are users supposed to test hardware AEC for device/OS version combinations in order to potentially get them whitelisted, when starting with Signal 4.3.0, software AEC is used when a device/OS combination is not on the whitelist already...?

**Comment by jeremymasters** [comment]
Another thought on this is that sometimes a junky driver can make it echo until you reboot and then it's fine for awhile, emphasizing RiseT's point of Phone+OS unique identifier for AEC style

**Comment by s1awa** [comment]
Just let users enable hardware AEC in the settings for better performance (with a warning that this causes echo on some devices). This option will be disabled by default.
If they experience echo they will know the culprit and can turn it off again.

**Comment by RiseT** [comment]
> Just let users enable hardware AEC in the settings for better performance (with a warning that this causes echo on some devices). This option will be disabled by default.

That's the only solution I can think of that doesn't require constant pampering.

**Comment by haffenloher** [comment]
> I've switched to a whitelist, so let's start this process over again. =) Any devices we know do work?

Two more models which worked fine before the switch to a whitelist: "D5803" (Sony Xperia Z3 Compact) and "FP1" (Fairphone 1).

**Comment by minnmann** [comment]
> Two more models which worked fine before the switch to a whitelist: "D5803" (Sony Xperia Z3 Compact) and "FP1" (Fairphone 1).

The same applies to: "SM-A500FU" (Samsung Galaxy A5 2015) and "XT1092" (Motorola Moto X 2014).
Full device names (from debug log) are: "samsung SM-A500FU (a5ultexx)" and "motorola XT1092 (victara_retde)".

We had many video calls with these devices and the audio quality was superb (indoors with WiFi and outdoors (only voice) with 4G). After the update to 4.4, we now have inferior quality (abrupt changes in volume etc.), even in cases, where there are absolutely no surrounding noises (e.g. both are at home). 

If you need any more information (e.g. a debug log) or if I should open a separate issue, just let me know!

**Comment by sigenc** [comment]
I also think with 7.1.2 Nexus 6P echo got fixed

**Comment by xenithorb** [comment]
Wanted to report back in reference to [my comment 1 month ago](https://github.com/WhisperSystems/Signal-Android/issues/6241#issuecomment-291709876) that bacon is now fixed either due to the whitelist or 7.1.2 (lineage). I do not know which but finally the non-proximity related screeching noises are gone and I can go back to using calling again.

**Comment by haffenloher** [comment]
> I also think with 7.1.2 Nexus 6P echo got fixed

@sigenc what does that mean? Did you test this with a version of Signal older than v4.3.0 (i.e. without 51f27631)?

**Comment by dcherian** [comment]
Moto G4 Plus (athene_f) has  worked well all the time.

**Comment by micressor** [comment]
For a better test: Is it possible to add `AEC` as a setting in signal?
I have two FP2 (Fairphone 2) on Android 5.1 with signal 4.5.3.
The audio quality depends on call to call excellent to inferior.
I'm not sure if it depends on the network quality.

If you need more info, just let me know.

**Comment by deano2390** [comment]
Just thought i'd drop a comment here because i'm facing similar issues in a closed source video chat application i'm working on.

In our app - if the call is between just 2 people we use webRTC. If there are more than 2 participants then we use our own A/V code and send it all through a server. When using our own code (which makes use of the Java layer audio classes AudioTrack and AudioRecord) we find that the hardware Echo cancellation works fine on all devices we can get hands on. But when the call utilises webRTC, certain devices hardware AEC fails completely despite working fine with the Java audio system. 

I have not yet dug too far into the webRTC codebase but it seems that there is something about the way the webRTC interfaces with the devices sound system that causes the hardware AEC to fail.

I notice you guys went down the route of whitelisting which is something i'm very eager to avoid!

I'll keep you posted if I manage to figure out exactly what webRTC is doing to cause problems with hardware AEC.

**Comment by moxie0** [comment]
@deano2390 FWIW the webrtc folks explicitly recommend disabling OpenSLES and always using the AudioTrack/AudioRecord interface (`WebRtcAudioManager.setBlacklistDeviceForOpenSLESUsage(true)`). 

The big remaining mystery to me now is how apps like Duo/Hangouts manage to use webrtc and actually ring correctly. My calls to AudioManager to setMode() for ringtone always get clobbered by webrtc.

**Comment by moxie0** [comment]
In the 4.7.1 beta, I've switched back to a blacklist. Here is my set of assumptions:

1) Hardware AEC seems to function properly when using the Java AudioTrack/AudioRecord interface (we think!). It seems to function poorly when using the OpenSL ES interface (causing this issue).
2) WebRTC AEC seems to function properly when using the OpenSL ES interface, but poorly when using the AudioTrack/AudioRecord interface. This causes issue #6737.
3) Many devices seem to "drop" audio after some period of time when using the OpenSL ES interface. This causes issue #6432.

As a result, there are three issues that are somewhat in tension with each-other: this issue, #6737, and #6432.

My current thinking is that we disable OpenSL ES by default to address #6432, that we leave hardware AEC on to address #6737, and that hopefully there aren't regressions with echo in this issue because we're using the AudioTrack/AudioRecord interface instead of OpenSL ES.

Please let me know if 4.7.1 causes regressions for anyone here (remember it would have to be the person on the other end of the call running 4.7.1 with a regression that would cause you to hear echo).

**Comment by FeuRenard** [comment]
I'm running 4.7.3 on Fairphone 2 (Android 6.0.1) and just got the feedback from a Signal-iOS user that she can hear her own echo.

**Comment by lantsem** [comment]
Moto X (2014) running LineageOS 14.1 got echo again with 4.7.4.

**Comment by haffenloher** [comment]
@lantsem What does that mean? _You're_ hearing echo when you're talking to _someone else_ who has a Moto X (2014) running LineageOS 14.1 and Signal 4.7.4?

**Comment by minnmann** [comment]
> Moto X (2014) running LineageOS 14.1 got echo again with 4.7.4.

I'm running 4.7.4 with Android 6.0 on a Moto X 2014 (XT1092), too (since the beta has been available). The other side is usually a Galaxy A5 2015 (SM-A500FU) and we do not have any echo or other calling issues, even outside when I'm using my Moto X and it's quite windy.

**Comment by lantsem** [comment]
@haffenloher Yes, from several phones that I tested with 4.7.4 only Moto X 2014 running LOS 14.1 shows an issue. But the impression was created that the echo became less noticeable for normal talk (not on speaker).

@minnmann My second Moto X 2014 with stock 6.0 didn't have echo with hardware AEC even with OpenSL ES. I wrote about it earlier.

**Comment by lantsem** [comment]
Moto G5 with stock 7.0 and 4.7.4 also has an echo with speaker mode.

**Comment by adlererik** [comment]
Was having echo issues with the Moto G5. Compiled v4.7.4 with add("Moto G5"); in the blacklist. 
No more tortuous echo for the people that listen to me. 

src/org/thoughtcrime/securesms/ApplicationContext.java

  private void initializeWebRtc() {
    Set<String> HARDWARE_AEC_BLACKLIST = new HashSet<String>() {{
      add("Pixel");
      add("Pixel XL");
      add("Moto G5");
......

Echo is gone. Works well now. Could you add the Moto G5. It has a very broken AEC

**Comment by haffenloher** [comment]
@adlererik with the Moto G5's stock rom or with a custom OS?

**Comment by moxie0** [comment]
@adlererik What happens when you *also* add Moto G5 to the opensles whitelist?

**Comment by adlererik** [comment]
@moxie0 Sorry for the late response. Have tested two Moto G5 phones with opensles whitelisted 

Signal v4.9.9 and v4.10.10 tested.  
Phone 1: Moto G5 100% stock unrooted 
Phone 2: Moto G5 LineageOS 

src/org/thoughtcrime/securesms/ApplicationContext.java

  private void initializeWebRtc() {
    try {
      Set<String> HARDWARE_AEC_BLACKLIST = new HashSet<String>() {{
        add("Pixel");
        add("Pixel XL");
        add("Moto G5");
      }};

      Set<String> OPEN_SL_ES_WHITELIST = new HashSet<String>() {{
        add("Pixel");
        add("Pixel XL");
        add("Moto G5");
      }};

Test conditions: 
Both Moto G5s called each other. Both Moto G5s called an iPhone 6.

Results: 
Additionally whitelisting  opensles ever so slightly *might* have improved the sound on the stock G5 during phone conversations. This is subjective. No improvement was noted with LineageOS.

Conclusion: 
a) Using AEC_BLACKLIST is a must have as AEC is broken on the G5.  OPEN_SL_ES_WHITELIST is probably redundant. 
b) Lenovo Moto hw is junk and their code is profoundly embarrassing.

**Comment by equaeghe** [comment]
My Samsung GT-I9192 Galaxy S4 Mini Duos (serranodsxx) with Signal 4.11.5 on Android 7.1.2 (LineageOS 14.1-20170615-nightly-serranodsdd) is causing echo for other end (Samsung Note 3, I think).

This echo gets *worse* when a headphone is used on my end. This means that the echo does not originate from my microphone picking up the other end's audio, but must happen inside the system/software stack. (Perhaps echo cancellation is effectively adding the echo?)

- Is it possible to add the GT-I9192 to the blacklist? (Cf. blacklist request for variant GT-I9195 above by @haffenloher.)
- Is it possible to disable echo cancellation when headphone output only is used? (Should be most energy efficient.)

The other end is my mother, who I'm trying to convince to continue using Signal, so this is important. ;-)

**Comment by haffenloher** [comment]
I don't think I ever encountered a Lineage OS phone which *didn't* produce echo. If it's their bug, putting every Lineage user's phone on Signal's hardware blacklist would be counterproductive.

**Comment by AlfonsoMuskedunder** [comment]
@haffenloher I have made Signal calls regularly on a Moto G, Moto G2, and Moto G4 Play, all on LineageOS, and never experienced echo issues.

**Comment by equaeghe** [comment]
@AlfonsoMuskedunder To be sure: you do mean that no echo was produced at the other end, no? In any case, it may still be a LineageOS bug, but just not for all devices.

@haffenloher 
Searching for  ‘echo’ in the LineagOS bug tracker, I find the following reports (all reported for specific devices) that may be relevant:

Most promising:
 
* [Other Persons hears echo on line and skype](https://jira.lineageos.org/browse/BUGBASH-827) – Also mentions echo is present when headphone is used
* [Phone echo](https://jira.lineageos.org/browse/BUGBASH-731) – Little information in bug report, but a [linked patch](https://review.lineageos.org/#/c/166653/) mentions something interesting about the second microphone:

  > default backends of IN_HDMI_MIC and IN_HANDSET_MIC by qualcomm are defined for single mic devices which breaks secondary mic for noise cancellation on dual mic devices like onyx while recording audio on apps like whatsapp, telegram etc.

May be useful:

* [Phone App Echo](https://jira.lineageos.org/browse/REGRESSION-263) – Seems to be about when on speaker with standard phone app
* [AEC (Acoustic Echo Cancellation) non-existent on calls in Hangouts or FreedomPop Messaging](https://jira.lineageos.org/browse/BUGBASH-705) – Problem disappears when headphones are used
* [Echo Using Google Hangouts](https://jira.lineageos.org/browse/BUGBASH-758) – Report specifically about when speaker is used

In any case, it may be useful to get a LineageOS developer to chip in here. I guess the best way is via a bug report in their tracker referencing this issue. I could do that, but I guess a Signal contributor would be better suited.

Also, to not let this idea fall by the wayside: Is disabling echo cancellation when using a headphone a possibility?

**Comment by lantsem** [comment]
@haffenloher Tested voice calls with Moto G and Galaxy S5, both on latest LineageOS 14.1 builds. They didn't have echo issue.

**Comment by Le1b1** [comment]
I have to report a massive echo during video calls with a Sony Xperia Z5 compact.

Model: E5823
Android-Version: 7.1.1
Signal-Version: 4.13.6 

The owner is willing to update to a test-build or join the Beta-club.

@McLoo Do you still use that model and people you talk to can confirm the problem?

**Comment by duckduckhacker** [comment]
I have an FP2 (Open OS v17.11.2) with Signal v4.13.6 installed via YalpStore. People I call experience an echo of themselves at their end unless I use headphones. This has been the case since August (following stable update releases).

**Comment by billie80** [comment]
If you read this thread from the beginning you will find out that this issue was originally about the FP2. Unfortunately this seems to be an issue with the hardware AEC of the Fairphone ROM. Additionally working around this by blacklisting hw AEC seems to make the phone crash. So if I got it correct Fairphone should either fix their hardware AEC or the crash when it is blacklisted.

**Comment by janvlug** [comment]
I experience an echo with someone who is using the following:
Device  : OnePlus ONE E1003 (OnePlus)
Android : 5.1.1 (4, ONE E1003_14_160420)
Memory  : 38M (33.31% free, 512M max)
Memclass: 192
OS Host : ubuntu-56
App : Signal 4.14.10

**Comment by ghostofkendo** [comment]
Similarly to @ghost, I have a Fairphone 2 running Fairphone Open OS v17.12.1 (based on Android 6.0.1) and Signal v4.14.10.
The people I call hear an echo of themselves as long as I don't use headphones, but I don't hear an echo of myself.

So I tried adding the Fairphone 2 to Signal's hardware AEC blacklist as follows, in `src/org/thoughtcrime/securesms/ApplicationContext.java`:
```
private void initializeWebRtc() {
    try {
      Set<String> HARDWARE_AEC_BLACKLIST = new HashSet<String>() {{
        add("Pixel");
        add("Pixel XL");
        add("Moto G5");
        add("FP2");
      }};
```
I only tried this change with a single call for now, but the other party's echo was almost gone (i.e. it was still there, but its volume was so low it wasn't really bothersome) and my FP2 did not crash.
The only downside is that I could hear a low-volume echo of myself, just like the other party, during that call.
Anyway, the situation is still better than with Signal 4.14.10 vanilla.

**Comment by minnmann** [comment]
After updating to 4.15 (beta), the following combination leads to an echo when using video chat:
Me: "motorola Moto G (5) (cedric)" (Moto G5), Signal 4.15 (beta)
Other person: "SM-A500FU" (Samsung Galaxy A5 2015), Signal 4.14 as well as Signal 4.15 (beta)

The echo occurs only on the other phone (Samsung A5). After 5-10 minutes the echo vanishes; only then and now some minor noise can be heard.

**Comment by codingluke** [comment]
I really appreciate the effort and all the new features coming to signal last year!

However, I can support @Le1b1 on echoing problems with the `Sony Xperia Z5 Compact`. I can hear my counterpart clearly, but they experience bad quality since the change to webRTC. This makes my live bringing friends and family to Signal quite hard. It would be very interesting to see what happens by adding all the Xperia Z5 models to the blacklist, as I only use signal for messaging these days. 

I also support a settings feature to Enable/Disable AEC. I see the point that it is not easy understandable for non-techies that the one with the clear audio has to disable it. So what about a dialog after the call which asks something like, "did **your counterpart** complained about hearing himself while the call? (yes/no)" when they click yes, It will disable it, so there is no technical understanding needed.

```
Name: Sony Xperia Z5 Compact
Model: E5803
Android-Version: 7.1.1
Signal-Version: 4.15.5 / Happens also on all minor versions since webRTC
```

**Comment by mkampl** [comment]
My communication partners also hear them self, if I use headphones or speaker.

Phone: Honor 5X (KIW-L21)
OS: LineageOS 14.1 (Android 7.1.2)
Signal: 4.15.5

**Comment by xander9285** [comment]
There's a strong echo with Motorola Moto G 3rd generation (XT1540), stock Android 6.0, current Signal version 4.15.5 .
I've had to switch to Wire's videochat, which isn't as good.  Signal's current blacklist approach is not sustainable.

**Comment by ADepic** [comment]
I have a oneplus 2 running lineage OS 14.1 and every time I call my friend on speakerphone he (has a  iPhone 5 or something) complains how he can hear himself louder than me.

This means I can't use speaker phone.

Signal version: 4.15.5 ( had this issues a few versions back too... Since I started using Signal which was 5 months ago)

**Comment by MiscDev** [comment]
The Moto G4, running LineageOS 14.1, also causes the person on the other end to hear an echo when speaker is enabled.
```
[ro.product.device]: [athene]
[ro.product.manufacturer]: [Motorola]
[ro.product.model]: [Moto G4]
[ro.product.name]: [lineage_athene]
```

**Comment by automated-signal** [comment]
GitHub Issue Cleanup:
See #7598 for more information.
